File: labs/control-ui/api-shadow/on_device_dpi_evader.ts
// Control UI TypeScript: On-Device DPI Evasion API

export class OnDeviceDpiEvader {
  public defaultSplitOffset: number;
  public evadedCount: number = 0;

  constructor(defaultSplitOffset: number = 2) {
    this.defaultSplitOffset = defaultSplitOffset;
  }

  public applySniSplit(clientHello: Uint8Array, splitOffset: number): Uint8Array[] {
    this.evadedCount++;
    if (clientHello.length < 5 || clientHello[0] !== 0x16) {
      return [clientHello];
    }
    const pos = Math.max(1, Math.min(splitOffset, clientHello.length - 1));
    return [clientHello.slice(0, pos), clientHello.slice(pos)];
  }

  public applyHttpDesync(request: Uint8Array): Uint8Array {
    this.evadedCount++;
    const str = new TextDecoder().decode(request);
    const lines = str.split('\r\n');
    const desynced = lines.map((line) => {
      if (line.toLowerCase().startsWith('host:')) {
        const parts = line.split(':');
        if (parts.length >= 2) {
          return `hOst: ${parts.slice(1).join(':').trim()}`;
        }
      }
      return line;
    });
    return new TextEncoder().encode(desynced.join('\r\n'));
  }

  public generateOutOfOrderChunks(data: Uint8Array, chunkSize: number): Uint8Array[] {
    this.evadedCount++;
    const size = Math.max(1, chunkSize);
    const chunks: Uint8Array[] = [];
    let offset = 0;

    while (offset < data.length) {
      const end = Math.min(offset + size, data.length);
      chunks.push(data.slice(offset, end));
      offset = end;
    }

    if (chunks.length > 1) {
      const temp = chunks[0];
      chunks[0] = chunks[1]!;
      chunks[1] = temp!;
    }
    return chunks;
  }

  public craftTtlDecoyPair(
    realPayload: Uint8Array,
    decoyTtl: number,
  ): {
    decoy: { ttl: number; payload: Uint8Array };
    real: { ttl: number; payload: Uint8Array };
  } {
    this.evadedCount++;
    const decoy = new Uint8Array(realPayload.length);
    for (let i = 0; i < realPayload.length; i++) {
      decoy[i] = realPayload[i]! ^ 0x55;
    }
    return {
      decoy: { ttl: decoyTtl, payload: decoy },
      real: { ttl: 64, payload: realPayload },
    };
  }
}

File: labs/mobile/donor-archive/tunnel/OnDeviceDpiEvader.kt
package com.luminet.android.tunnel

class OnDeviceDpiEvader(val defaultSplitOffset: Int = 2) {
    var evadedCount: Long = 0
        private set

    fun applySniSplit(clientHello: ByteArray, splitOffset: Int): List<ByteArray> {
        evadedCount++
        if (clientHello.size < 5 || clientHello[0] != 0x16.toByte()) {
            return listOf(clientHello)
        }
        val pos = splitOffset.coerceIn(1, clientHello.size - 1)
        return listOf(
            clientHello.copyOfRange(0, pos),
            clientHello.copyOfRange(pos, clientHello.size)
        )
    }

    fun applyHttpDesync(request: ByteArray): ByteArray {
        evadedCount++
        val str = try { String(request, Charsets.UTF_8) } catch (e: Exception) { return request }
        val lines = str.lines()
        val modified = lines.joinToString("\r\n") { line ->
            if (line.lowercase().startsWith("host:")) {
                val parts = line.split(":", limit = 2)
                if (parts.size == 2) "hOst: ${parts[1].trim()}" else line
            } else {
                line
            }
        }
        return modified.toByteArray(Charsets.UTF_8)
    }

    fun generateOutOfOrderChunks(data: ByteArray, chunkSize: Int): List<ByteArray> {
        evadedCount++
        val size = maxOf(1, chunkSize)
        val chunks = mutableListOf<ByteArray>()
        var offset = 0
        while (offset < data.size) {
            val end = minOf(offset + size, data.size)
            chunks.add(data.copyOfRange(offset, end))
            offset = end
        }

        if (chunks.size > 1) {
            val tmp = chunks[0]
            chunks[0] = chunks[1]
            chunks[1] = tmp
        }
        return chunks
    }

    fun craftTtlDecoyPair(realPayload: ByteArray, decoyTtl: Byte): Pair<Pair<Byte, ByteArray>, Pair<Byte, ByteArray>> {
        evadedCount++
        val decoy = realPayload.clone()
        for (i in decoy.indices) {
            decoy[i] = (decoy[i].toInt() xor 0x55).toByte()
        }
        return Pair(Pair(decoyTtl, decoy), Pair(64.toByte(), realPayload))
    }
}
